core/index/index_writer.cpp

```diff
@@ -48,108 +48,129 @@ constexpr size_t kNonUpdateRecord = std::numeric_limits<size_t>::max();
-    std::span<const index_writer::modification_context> modification_contexts)
-    : doc_id_begin_(doc_id_begin),
-      doc_id_end_(doc_id_end),
-      segment_(segment),
+    std::span<const IndexWriter::ModificationContext> modification_contexts)
+    : doc_id_begin_{doc_id_begin},
+      doc_id_end_{doc_id_end},
+      segment_{std::move(segment)},
+      docs_mask_{std::move(docs_mask)},
       modification_contexts_{modification_contexts},
-      update_contexts_{update_contexts} {
+      update_contexts_{update_contexts},
+      reader_{std::move(reader)} {
     IRS_ASSERT(doc_id_begin_ <= doc_id_end_);
     IRS_ASSERT(doc_id_end_ - doc_limits::min() <= segment_.meta.docs_count);
     IRS_ASSERT(update_contexts.size() == segment_.meta.docs_count);
   }
 };
 
-std::vector<index_file_refs::ref_t> extract_refs(
-  const ref_tracking_directory& dir) {
-  std::vector<index_file_refs::ref_t> refs;
-  // FIXME reserve
-
-  auto visitor = [&refs](const index_file_refs::ref_t& ref) {
-    refs.emplace_back(ref);
-    return true;
-  };
-  dir.visit_refs(visitor);
-
-  return refs;
-}
-
 // Apply any document removals based on filters in the segment.
 // modifications where to get document update_contexts from
 // docs_mask where to apply document removals to
 // readers readers by segment name
 // meta key used to get reader for the segment to evaluate
 // min_modification_generation smallest consider modification generation
 // Return if any new records were added (modification_queries_ modified).
-bool add_document_mask_modified_records(
-  std::span<index_writer::modification_context> modifications,
-  document_mask& docs_mask, readers_cache& readers, segment_meta& meta,
-  size_t min_modification_generation = 0) {
+void RemoveFromExistingSegment(
+  std::vector<doc_id_t>& deleted_docs,
+  std::span<IndexWriter::ModificationContext> modifications,
+  const SubReader& reader, size_t min_modification_generation = 0) {
+  IRS_ASSERT(deleted_docs.empty());
+
   if (modifications.empty()) {
-    return false;  // nothing new to flush
+    return;  // nothing new to flush
   }
 
-  auto reader = readers.emplace(meta);
+  auto& docs_mask = *reader.docs_mask();
+
+  for (auto& modification : modifications) {
+    if (IRS_UNLIKELY(!modification.filter)) {
+      continue;  // skip invalid or uncommitted modification queries
+    }
+
+    auto prepared = modification.filter->prepare(reader);
+
+    if (IRS_UNLIKELY(!prepared)) {
+      continue;  // skip invalid prepared filters
+    }
+
+    auto itr = prepared->execute(reader);
+
+    if (IRS_UNLIKELY(!itr)) {
+      continue;  // skip invalid iterators
+    }
+
+    while (itr->next()) {
+      const auto doc_id = itr->value();
+
+      // if the indexed doc_id was insert()ed after the request for modification
+      // or the indexed doc_id was already masked then it should be skipped
+      if (modification.generation < min_modification_generation ||
```

Comment:
modification.generation < min_modification_generation

This part of condition is always true/false for entire cycle

So I think better to check it before cycle and even before `(IRS_UNLIKELY(!modification.filter)) {`

Reply:
:+1: